### PR TITLE
Fix TLS decryption error

### LINKED ISSUE #153: Fix mstsc.exe SChannel decryption failure on client
During the first few seconds of the RDP connection, mstsc.exe with an injected sspi-rs fails to decrypt traffic and terminates:

image

image

It's unclear what the problem is, but it prevents a successful mstsc.exe connection using sspi-rs, and it happens every time for me.

### PR DESCRIPTION
**Pull request by TheBestTvarynka**
Hi,
In this pull request, I have fixed the decryption error. This error used to look like this:

image

If you think "_This pull request does not look like a fix_", then I can understand you 😅 

Actually, to fix all the problems I was forced to change the TLS decryption algorithm a lot, fix the memory leaks, change the API, and other things. I recommend you review it commit by commit. I made comprehensive commit messages with problem explanations and descriptions of each fix. Also, I would like to merge and rebase instead of merge and squash this pull request. It would be great to save those explanations in the commit history.

The proposed changes contain some undocumented anywhere behavior. I added comments where appropriate. Feel free to ask any questions.

closes #153

### CODE REVIEW

File: src/credssp/mod.rs [hunk 1]
```diff
@@ -734,6 +749,21 @@ impl<'a> SspiContext {
...
+        Generator::new(move |mut yield_point| async move {
+            self.initialize_security_context_impl(&mut yield_point, builder).await
+        })
+        .resolve_with_default_network_client()
+    }
+
+    pub fn change_password_sync(&mut self, builder: ChangePassword) -> crate::Result<()> {
+        Generator::new(move |mut yield_point| async move { self.change_password_impl(&mut yield_point, builder).await })
+            .resolve_with_default_network_client()
+    }
```
Comment (TheBestTvarynka) [inline src/credssp/mod.rs:765]: We need those two functions to eliminate lifetime errors in the FFI.

File: src/credssp/sspi_cred_ssp/tls_connection.rs [hunk 2]
```diff
@@ -63,21 +136,152 @@ impl TlsConnection {
...
-                    .process_new_packets()
-                    .map_err(|err| Error::new(ErrorKind::DecryptFailure, err.to_string()))?;
+                while !tls_packet.is_empty() {
+                    let _ = tls_connection.read_tls(&mut tls_packet)?;
 
-                let mut reader = tls_connection.reader();
+                    let tls_state = tls_connection
+                        .process_new_packets()
+                        .map_err(|err| Error::new(ErrorKind::DecryptFailure, err.to_string()))?;
+                    trace!(?tls_state);
```
Comment (TheBestTvarynka) [inline src/credssp/sspi_cred_ssp/tls_connection.rs:255]: nope, it's not a leftover 😃 
It's very convenient to see how many bytes the `rustls` has managed to decrypt. Also, it shows us an overall IO state.
Sometimes, it can be very useful during debugging.

File: src/decrypt_buffer.rs
Comment (TheBestTvarynka) [inline src/decrypt_buffer.rs:1]: In the scope of this pull request, I have fixed the memory leak during the decryption and made it fully in-place by using references. But we still have a similar error in the `EncryptMessage` function.
Fixing the `EncryptMessage` function is out of the scope of this PR and not critical (`mstsc` works well anyway). Shout I create an issue about it?

**Comment by CBenoit** [comment]
> Also, I would like to merge and rebase instead of merge and squash this pull request. It would be great to save those explanations in the commit history.

I agree with this. However, it would be best if each commit could pass the CI.
As a best effort, could you run this for each commit and amend as appropriate?

```bash
$ cargo clippy --workspace -- -D warnings && cargo test --workspace && pwsh ./tools/wasm-testcompile/check.ps1 && cargo fmt --check
```

Here is how you can do this automatically for each commit:
```bash
$ git rebase -i --exec 'cargo clippy --workspace -- -D warnings && cargo test --workspace && pwsh ./tools/wasm-testcompile/check.ps1 && cargo fmt --check' origin/master
```

On failure, the rebase will stop so that you can fix the problem and amend using `git commit --amend`.

Thank you!

File: Cargo.lock
Comment (CBenoit) [inline Cargo.lock:1]: question: Is it necessary to update the Cargo.lock as part of this PR? I would rather leave that to a separate PR (in fact, we should enable dependabot) and remove "noise" from this PR. By noise I mean that if something breaks because of a dependency, I would rather have a commit dedicated to dependencies.

**Comment by TheBestTvarynka** [comment]
> could you run this for each commit and amend as appropriate?

No problem. I'll do it

**Comment by TheBestTvarynka** [comment]
> Is it necessary to update the Cargo.lock as part of this PR?

Nope. I just played with different `rustls` versions. I'll revert it

**Comment by TheBestTvarynka** [comment]
@CBenoit I think you can review it in the meantime when I'm fixing the CI

File: src/credssp/sspi_cred_ssp/tls_connection.rs [hunk 3]
```diff
@@ -66,17 +66,21 @@ impl TlsConnection {
     pub fn decrypt_tls(&mut self, mut payload: &[u8]) -> Result<Vec<u8>> {
         match self {
             TlsConnection::Rustls(tls_connection) => {
-                tls_connection.read_tls(&mut payload)?;
+                let mut plain_data = Vec::with_capacity(payload.len());
 
-                let tls_state = tls_connection
-                    .process_new_packets()
-                    .map_err(|err| Error::new(ErrorKind::DecryptFailure, err.to_string()))?;
+                while payload.len() != 0 {
```
Comment (CBenoit) [inline src/credssp/sspi_cred_ssp/tls_connection.rs]: suggestion: `while !payload.is_empty()`

File: src/credssp/sspi_cred_ssp/tls_connection.rs [hunk 4]
```diff
@@ -66,17 +66,21 @@ impl TlsConnection {
...
+                        .process_new_packets()
+                        .map_err(|err| Error::new(ErrorKind::DecryptFailure, err.to_string()))?;
 
-                let mut plain_data = vec![0; tls_state.plaintext_bytes_to_read()];
-                let _plain_data_len = reader.read(&mut plain_data)?;
+                    let mut reader = tls_connection.reader();
+                    let mut decrypted = vec![0; tls_state.plaintext_bytes_to_read()];
+                    let _plain_data_len = reader.read(&mut decrypted)?;
 
+                    plain_data.extend_from_slice(&decrypted);
```
Comment (CBenoit) [inline src/credssp/sspi_cred_ssp/tls_connection.rs]: suggestion: Instead of allocating an intermediate buffer each time, maybe you could use [`read_to_end`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#method.read_to_end).

```rust
reader.read_to_end(&mut plain_data)?;
```

It’s also less error prone that ignoring the read length returned by `read`. (I assume the current code is not broken, but typically we should check this value.)

**Review by CBenoit — COMMENTED** [review]
nitpick: It would be better to not include periods in the commit message. It’s not useful and increases chances of getting the message wrapped (too long). Of course, it’s not a problem in the body of the commit.

File: src/decrypt_buffer.rs
Comment (CBenoit) [reply src/decrypt_buffer.rs:1]: Yes please!

File: Cargo.toml [hunk 5]
```diff
@@ -64,7 +64,8 @@ trust-dns-resolver = { version = "0.23", optional = true }
 portpicker = { version = "0.1", optional = true }
 num-bigint-dig = "0.8"
 tracing = "0.1"
-rustls = { version = "0.21", features = ["dangerous_configuration"], optional = true }
+rustls = { path = "../deleteIt/rustls/rustls", optional = true }
+#rustls = { version = "0.22.2", optional = true }
```
Comment (CBenoit) [inline Cargo.toml]: issue: This intermediate commit is referring to a local fork. This commit should be amended to pass the CI if we want to rebase.

File: ffi/src/sspi/common.rs [hunk 6]
```diff
@@ -265,6 +275,7 @@ pub unsafe extern "system" fn EncryptMessage(
     p_message: PSecBufferDesc,
     message_seq_no: u32,
 ) -> SecurityStatus {
+    info!("threadid: {:?} {:?}", std::thread::current().id(), *ph_context);
```
Comment (CBenoit) [inline ffi/src/sspi/common.rs]: question: Do we really need INFO-level traces of the current thread ID?

File: src/credssp/sspi_cred_ssp/tls_connection.rs [hunk 7]
```diff
@@ -117,20 +153,21 @@ impl TlsConnection {
...
+                //     rustls::SupportedCipherSuite::Tls12(cipher_suite) => &cipher_suite.common.bulk,
+                //     rustls::SupportedCipherSuite::Tls13(cipher_suite) => &cipher_suite.common.bulk,
+                // };
+                // let block_size = match bulk_cipher {
+                //     rustls::BulkAlgorithm::Aes128Gcm => AES_BLOCK_SIZE,
+                //     rustls::BulkAlgorithm::Aes256Gcm => AES_BLOCK_SIZE,
+                //     // ChaCha20 is a stream cipher
+                //     rustls::BulkAlgorithm::Chacha20Poly1305 => 0,
+                // };
+                let block_size = AES_BLOCK_SIZE;
```
Comment (CBenoit) [inline src/credssp/sspi_cred_ssp/tls_connection.rs]: issue: This commit seems to include many changes unrelated to the Mutex addition

File: src/builders/accept_sec_context.rs [hunk 8]
```diff
@@ -117,15 +98,13 @@ impl<
...
         'a,
-        AuthData,
         CredsHandle,
         WithCredentialsHandle,
         ContextRequirementsSet,
         TargetDataRepresentationSet,
         OutputSet,
     > {
         AcceptSecurityContext {
-            inner: self.inner,
```
[the remark this answers is not in the document]
Comment (CBenoit) [reply src/builders/accept_sec_context.rs:128]: This may be because the lifetime bounds are not properly specified.

**Comment by CBenoit** [comment]
note: I see that [this commit](https://github.com/Devolutions/sspi-rs/pull/216/commits/d57b886d845fb8c7e7d2ade670f7d943c0e8c4cb) is including two different changes and both are specified using a conventional commit message, but both are included in the same commit:

image

FYI, this is not following the specification. This should be two different commits. This is defeating the purpose of being machine readable. The body may contain an extended description of the change of the commit, and optional footers with additional machine-readable metadata, but not another commit message.

File: src/credssp/mod.rs [hunk 9]
```diff
@@ -502,18 +513,22 @@ impl<C: CredentialsProxy<AuthenticationData = AuthIdentity>> CredSspServer<C> {
...
-                        .with_output(&mut output_token)
-                        .execute(),
+                    AcceptSecurityContext::<
+                        '_,
+                        _,
+                        WithoutCredentialsHandle,
+                        WithoutContextRequirements,
+                        WithoutTargetDataRepresentation,
+                        WithoutOutput,
+                    >::new()
```
Comment (CBenoit) [inline src/credssp/mod.rs:525]: thought: Not very ergonomic. I would expect that we don’t need to specify all the generic parameters.

**Comment by CBenoit** [comment]
We can try to proceed with the rebase approach this time, but I think that it would be generally simpler to just open several PRs to avoid merging intermediate commits that do not pass the CI. Feel free to open many small PRs at any time. The smaller, the quicker I’ll review and merge it.

File: src/credssp/sspi_cred_ssp/tls_connection.rs (hunk 2 above)
Comment (TheBestTvarynka) [reply src/credssp/sspi_cred_ssp/tls_connection.rs:255]: done: https://github.com/Devolutions/sspi-rs/issues/217 

I'll link to this issue example of the solution for the decryption problem once the corresponding PR is merged.

File: src/decrypt_buffer.rs
Comment (TheBestTvarynka) [reply src/decrypt_buffer.rs:1]: done: https://github.com/Devolutions/sspi-rs/issues/217 

I'll link to this issue example of the solution for the decryption problem once the corresponding PR is merged.
